sound/soc/sof/pm.c

```diff
@@ -265,14 +265,17 @@ static int sof_suspend(struct device *dev, bool runtime_suspend)
 			ret);
 
 	/* Do not reset FW state if DSP is in D0 */
-	if (target_state == SOF_DSP_PM_D0)
-		return ret;
+	if (ret == 0 && target_state == SOF_DSP_PM_D0)
+		return 0;
 
 	/* reset FW state */
 	sof_set_fw_state(sdev, SOF_FW_BOOT_NOT_STARTED);
 	sdev->enabled_cores_mask = 0;
 
-	return ret;
+	/* Don't report failure!  It aborts the system suspend, and we
+	 * will recover naturally on resume anyway.
+	 */
+	return 0;
```

Comment:
we may need additional state changes to prevent the use of IMR on the next resume if the suspend failed, so that you really start from a clean slate. 

@ujfalusi added a new state for this, if the suspend is not successful maybe we need to set the state to FW_CRASHED. see "ASoC: SOF: Introduce new firmware state: SOF_FW_CRASHED"

Reply:
There's also an additional issue that when we suspend to S0idle aka S0i1 with some WoV stream active, we do not reset the DSP on resume, so the first change is not so good IMHO. If there is an issue when entering S0i1, is it really a problem to prevent that suspend and remain in S0?

Reply:
Not sure I understand?  The host can't touch IMR except via HDA DMA to the ROM firmware, and that part is definitely working.  The rig I was using[1] would just log a countdown via trace and issue an "ILL" instruction after 23 seconds[2].  The panic would get reported to the kernel and then the DSP would be dead.  Then after suspend (with this patch) I'd see the driver reload the firmware from scratch, which would start counting down again and blow up.  So nothing in the kernel seems to be seriously confused about device state, at least.

But yes: if there's a harder reset state, we should use that.  I'll see if that does anything interesting.

[1] On pre-Zephyr firmware!  I'm seriously out of my depth here.

[2] Experimentally enough time for me to get the device into a browser and start a video.

Reply:
Ah yes, thanks for pointing out that patch @plbossart adding a crashed state will indeed allow us to try and force the DSP to do things when we know its dead.

Regarding WoV, if DSP is dead, we can't open the stream so that should be a non issue.

Reply:
sure it's an issue. You have no way of restarting any sound functionality without a power cycle.

Reply:
Yes but we shouldnt kill the users device without warning by failing a suspend